abi.md

```diff
@@ -0,0 +1,164 @@
+---
+author:
+- The FIRRTL Specification Contributors
+title: Specification for the FIRRTL ABI
+revisionHistoryAbi: true
+---
+
+# Introduction
+
+FIRRTL defines a language/IR for describing synchronous hardware circuits.  This
+document specifies the mapping of FIRRTL constructs to Verilog in a manner
+similar to an application binary interface (ABI) which enables predictability of
+the output of key constructs necessary for the interoperability between circuits
+described in FIRRTL and between other languages and FIRRTL output.
+
+This document describes multiple versions of the ABI, specifically calling
+specific changes in later versions.  It is expected that a conforming FIRRTL
+compiler can lower to all specified ABIs.  This mechanism exists to allow
+improved representations when using tools which have better Verilog support and
+allow incremental migration of existing development flows to the significant 
+representational changes introduced by ABI changes.
+
+# FIRRTL System Verilog Interface
+
+To use a circuit described in FIRRTL in a predictable way, the mapping of certain
+behaviors and boundary constructs in FIRRTL to System Verilog must be defined.
+Where possible this ABI does not impose constraints on implementation,
+concerning itself primarily with the boundaries of the circuit.
+
+Two ABIs are defined.  ABIv1 describes a system in which no aggregate types
+appear on circuit boundaries (publicly visible locations).  This ABI captures
+and formalizes the historic behavior of FIRRTL to verilog lowering.  
+Specifically, this ABI, by construction, produces no aggregate Verilog types.
+ABIv2 defines an ABI in which aggregate types are preserved.  Both these ABIs 
+may evolve as new FIRRTL constructs are added.
+
+## On Modules
+
+### The Circuit and Top Level
+
+The top level module, as specified by the circuit name, shall be present as a
+System Verilog module of the same name.  This FIRRTL module is considered a
+"public" module and subject to the lowering constraints for public modules.
+
+### External Modules
+
+An external module may be presumed to exist following the lowering constraints
+for public modules.  The module shall exist with a verilog name matching the 
+defname value or, lacking that, the module name.
+
+###  Public Modules
+
+Any module considered a "public" module shall be implemented in Verilog in a
+consistent way.  Any public module shall exist as a Verilog module of the same
+name.
+
+Each public module with definitions (e.g. not external modules) shall be placed
+in a file with the same name.  No assumption is made of the filename of an
+implementation of an external module, it is the user's job to include files in
+such a way in their tools to resolve the name.
```

Comment:
This requirement of public modules being in separate files is reasonable. However, this has some complications:

1. Where is the sub-hierarchy of a public module emitted? What happens if two public modules instantiate the same private sub-module?
2. This disallows any single-file emission mode for any circuit which has a public module other than the just the main module. That's possibly fine, but may not be desired.

Reply:
public non-top modules is a new thing, so I don't think we are too worried about it.  Single-public top still supports single file emission.

as for (1), the more general question is "what are all the things I need to include and in what order?" and I've side stepped that a few places for now.